lib/cppcheck.cpp

```diff
@@ -1123,75 +1125,82 @@ unsigned int CppCheck::checkFile(const FileWithDetails& file, const std::string
-                }
+                    // dump xml if --dump
+                    if ((mSettings.dump || !mSettings.addons.empty()) && fdump.is_open()) {
+                        fdump << "<dump cfg=\"" << ErrorLogger::toxml(mCurrentConfig) << "\">" << std::endl;
+                        fdump << "  <standards>" << std::endl;
+                        fdump << "    <c version=\"" << mSettings.standards.getC() << "\"/>" << std::endl;
+                        fdump << "    <cpp version=\"" << mSettings.standards.getCPP() << "\"/>" << std::endl;
+                        fdump << "  </standards>" << std::endl;
+                        fdump << getLibraryDumpData();
+                        preprocessor.dump(fdump);
+                        tokenizer.dump(fdump);
+                        fdump << "</dump>" << std::endl;
+                    }
 
-                if (mSettings.inlineSuppressions) {
-                    // Need to call this even if the hash will skip this configuration
-                    mSuppressions.nomsg.markUnmatchedInlineSuppressionsAsChecked(tokenizer);
-                }
+                    if (mSettings.inlineSuppressions) {
+                        // Need to call this even if the hash will skip this configuration
+                        mSuppressions.nomsg.markUnmatchedInlineSuppressionsAsChecked(tokenizer);
+                    }
 
-                // Skip if we already met the same simplified token list
-                if (mSettings.force || mSettings.maxConfigs > 1) {
-                    const std::size_t hash = tokenizer.list.calculateHash();
-                    if (hashes.find(hash) != hashes.end()) {
-                        if (mSettings.debugwarnings)
-                            purgedConfigurationMessage(file.spath(), mCurrentConfig);
-                        continue;
+                    // Skip if we already met the same simplified token list
+                    if (mSettings.force || mSettings.maxConfigs > 1) {
+                        const std::size_t hash = tokenizer.list.calculateHash();
+                        if (hashes.find(hash) != hashes.end()) {
+                            if (mSettings.debugwarnings)
+                                purgedConfigurationMessage(file.spath(), mCurrentConfig);
+                            continue;
+                        }
+                        hashes.insert(hash);
                     }
-                    hashes.insert(hash);
-                }
 
-                // Check normal tokens
-                checkNormalTokens(tokenizer, analyzerInformation.get());
+                    // Check normal tokens
+                    checkNormalTokens(tokenizer, analyzerInformation.get());
+                } catch (const InternalError &e) {
+                    ErrorMessage errmsg = ErrorMessage::fromInternalError(e, &tokenizer.list, file.spath());
+                    mErrorLogger.reportErr(errmsg);
+                }
```

Comment:
The multiple layers of exception handling seem excessive. I think it is possible to clean this up a bit but that requires test coverage first.